src/codegen/program.lisp

```diff
@@ -41,41 +42,96 @@ A function bound here will be called with a keyword category, and one or more ad
 
     Toplevel definitions, after type checking and before compilation.")
 
+;; The following functions control the output order of compiled
+;; definitions and interleaved lisp expressions.
+;;
+;; Toplevel define and instance forms are compiled to 1 or more named,
+;; lisp-source-valued output definitions: when these definitions are
+;; generated, they are associated with the starting source offset of their
+;; toplevel form:
+;;
+;;   toplevel definition:
+;;     #<def .... (300 . 345)>
+;;
+;;   bindings (lisp definitions):
+;;     (300 b1 .. bn)
+;;
+;; Then when compile-definitions emits the full set of output
+;; definitions, any lisp source forms that occurred earlier in the
+;; file are emitted first.
+
+(defun bindings-offset (bindings offsets)
+  "Given a list of binding names, and a name -> offset map, return the earliest binding start offset."
+  (apply #'min
+         (mapcar (lambda (binding)
+                   (or (gethash (car binding) offsets) 0))
+                 bindings)))
+
+(defun merge-forms (lisp defs)
+  "Merge a list of toplevel-lisp forms and compiled definitions, ensuring that lisp forms are always placed before the original source locations of the compiled definitions."
```

Comment:
Some structural description of the arguments would be appreciated. 

E.g. ```LISP is a list of lists, (OFFSET SOMETHING-ELSE SOMETHING-ELSE-ELSE ...), and DEFS is ....```

or whatever.

Reply:
done